plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/purity/PurityChecker.kt

```
@@ -0,0 +1,180 @@
+            val isSelfCall = (exp.method.name == currentEvaluatedFunction?.name)
+
+            if (isSelfCall) {
+                exp.args.all(::isExpressionPure)
+            } else {
+                val mangledName = exp.method.name.mangled
+                val argsPure = exp.args.all(::isExpressionPure)
+                val operatorPure =
+                    PureOperatorWhitelist.isWhitelisted(mangledName) && argsPure
+                if (operatorPure) true
+                else {
+                    val callee = registry[exp.method.name]
+                    when (callee) {
+                        null -> {
+                            registry[exp.method.name] = PurityStatus.UNKNOWN
+                            addFunctionalDependency(exp.method.name)
+                            registry[currentEvaluatedFunction!!.name!!] = PurityStatus.CHECKING
+                            false
+                        }
+                        PurityStatus.CHECKING -> false
+                        PurityStatus.PURE -> true
+                        PurityStatus.IMPURE -> false
+                        else -> false
+                    }
+                }
+            }
+        }
+
+        // Single-expression If - just allow this for now
+        is If -> isExpressionPure(exp.condition) &&
+                isExpressionPure(exp.thenBranch) &&
+                isExpressionPure(exp.elseBranch)
+
+        is Goto -> true          // auto-generated return
+        is InhaleDirect -> true  // auto-generated check
+
+        // Local Variable Declarations
+        is Declare -> false
+        else -> false
+    }
+
+    private fun addFunctionalDependency(callee: MangledName) {
+        waiting.getOrPut(callee) { mutableSetOf() }
+            .add(currentEvaluatedFunction!!)
+
+        dbg { "${currentEvaluatedFunction!!.name!!.mangled} is now waiting on ${callee.mangled} (dependency not yet resolved)" }
+    }
```

Except for inline functions, it generally doesn't make sense to check purity transitively: we only want to raise the error "impure function annotated pure" once.